### DISCUSSION TITLE
Explicit Interface implementation instead of Virtual Method is a bug or a feature?

### DISCUSSION THREAD

**Discussion by ackava** [dotnet/csharplang#3996] [Q&A]
I accidentally implemented interface in multiple level and I found something unexpected.

```c#
using System;
					
public class Program
{
	public static void Main()
	{
		IShape s = new Shape();
		IShape c = new Circle();

		Console.WriteLine(Display(s, 1,2));
		Console.WriteLine(Display(c, 1,2));
	}
	
	public static int Display(IShape s, int a, int b) {
		return s.Do(a, b);
	}
}

public class Shape: IShape {
	int IShape.Do(int a, int b) {
	    return a + b;
	}
}

public class Circle: Shape, IShape {
	int IShape.Do(int a, int b) {
	    return a * b;
	}
}

public interface IShape {
	int Do(int a, int b);
}
```

Circle shape inherits Shape, but implements IShape again explicitly. And when I call `Do` on `Circle`, it calls `IShape.Do` of circle and not of the base class (it does not have `new` keyword, also it doesn't complain that it should have `new` keyword) . It outputs 3 for Shape and 2 for Circle.

I found this interesting as I can use it instead of virtual method. Not only this, I can explicitly override base method functionality even if it is not virtual provided caller uses only interface.

It turns out that explicit interface implementation is slightly slower,  https://dotnetfiddle.net/EGEz1m

**Comment by huoyaoyuan** [comment]
**Virtual methods** and **interface dispatch** are quite similar, but they are separated.

The effective `IShape.Do` implementation in derived class is the re-implemented one.
`new` keyword is a C# feature. In this case, the base implementation is un-callable in derived class, so no `new` keyword required.

**Comment by huoyaoyuan** [comment]
> It turns out that explicit interface implementation is slightly slower

Yes it does.

|    Method |           Job |       Runtime |     Mean |     Error |    StdDev | Ratio |
|---------- |-------------- |-------------- |---------:|----------:|----------:|------:|
|   Virtual |      .NET 4.8 |      .NET 4.8 | 2.259 ns | 0.0147 ns | 0.0130 ns |  1.03 |
| Interface |      .NET 4.8 |      .NET 4.8 | 2.963 ns | 0.0068 ns | 0.0061 ns |  1.35 |
|   Virtual | .NET Core 3.1 | .NET Core 3.1 | 2.175 ns | 0.0067 ns | 0.0052 ns |  0.99 |
| Interface | .NET Core 3.1 | .NET Core 3.1 | 2.737 ns | 0.0086 ns | 0.0072 ns |  1.25 |
|   Virtual | .NET Core 5.0 | .NET Core 5.0 | 2.187 ns | 0.0155 ns | 0.0145 ns |  1.00 |
| Interface | .NET Core 5.0 | .NET Core 5.0 | 2.755 ns | 0.0083 ns | 0.0074 ns |  1.26 |

The reason is the structure of internal CLR implementation. Virtual method info is direct in type info, and interface dispatch list is indirect.

The difference is so small and bounded to implementation detail, that you shouldn't consider it.

BTW please use [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet) to get reliable performance result.
